diff -Naur grep-3.2/configure.ac grep-3.2.tpg/configure.ac
--- grep-3.2/configure.ac	2018-12-16 02:35:51.000000000 +0000
+++ grep-3.2.tpg/configure.ac	2018-12-20 21:03:14.755941433 +0000
@@ -94,6 +94,18 @@
 AC_C_CONST
 gl_INIT
 
+AC_CACHE_CHECK([for gcc 7-compatible __builtin_mul_overflow_p],
+  [utils_cv_builtin_mul_overflow_p],
+  [AC_COMPILE_IFELSE(
+    [AC_LANG_PROGRAM(
+      [[]],
+      [[return __builtin_mul_overflow_p(1, 2, 3);]])],
+    [utils_cv_builtin_mul_overflow_p=yes],
+    [utils_cv_builtin_mul_overflow_p=no])])
+if test $utils_cv_builtin_mul_overflow_p = yes; then
+  AC_DEFINE([HAVE_BUILTIN_MUL_OVERFLOW_P], [1], [Define if your compiler supports __builtin_mul_overflow_p])
+fi
+
 # The test suite needs to know if we have a working perl.
 # FIXME: this is suboptimal.  Ideally, we would be able to call gl_PERL
 # with an ACTION-IF-NOT-FOUND argument ...
diff -Naur grep-3.2/gnulib-tests/macros.h grep-3.2.tpg/gnulib-tests/macros.h
--- grep-3.2/gnulib-tests/macros.h	2018-01-07 00:07:08.000000000 +0000
+++ grep-3.2.tpg/gnulib-tests/macros.h	2018-12-20 21:03:59.592940813 +0000
@@ -22,7 +22,7 @@
 #include <stdlib.h>
 
 #ifndef FALLTHROUGH
-# if __GNUC__ < 7
+# if __GNUC__ < 7 || !__has_attribute(__fallthrough__)
 #  define FALLTHROUGH ((void) 0)
 # else
 #  define FALLTHROUGH __attribute__ ((__fallthrough__))
diff -Naur grep-3.2/gnulib-tests/vasnprintf.c grep-3.2.tpg/gnulib-tests/vasnprintf.c
--- grep-3.2/gnulib-tests/vasnprintf.c	2018-10-16 23:22:22.000000000 +0000
+++ grep-3.2.tpg/gnulib-tests/vasnprintf.c	2018-12-20 21:04:38.031940282 +0000
@@ -119,7 +119,7 @@
 #endif
 
 #ifndef FALLTHROUGH
-# if __GNUC__ < 7
+# if __GNUC__ < 7 || !__has_attribute(__fallthrough__)
 #  define FALLTHROUGH ((void) 0)
 # else
 #  define FALLTHROUGH __attribute__ ((__fallthrough__))
diff -Naur grep-3.2/lib/dfa.c grep-3.2.tpg/lib/dfa.c
--- grep-3.2/lib/dfa.c	2018-12-15 18:14:26.000000000 +0000
+++ grep-3.2.tpg/lib/dfa.c	2018-12-20 21:05:07.032939881 +0000
@@ -55,7 +55,7 @@
 #include "localeinfo.h"
 
 #ifndef FALLTHROUGH
-# if __GNUC__ < 7
+# if __GNUC__ < 7 || !__has_attribute(__fallthrough__)
 #  define FALLTHROUGH ((void) 0)
 # else
 #  define FALLTHROUGH __attribute__ ((__fallthrough__))
diff -Naur grep-3.2/lib/fnmatch.c grep-3.2.tpg/lib/fnmatch.c
--- grep-3.2/lib/fnmatch.c	2018-01-07 00:08:02.000000000 +0000
+++ grep-3.2.tpg/lib/fnmatch.c	2018-12-20 21:05:37.521939459 +0000
@@ -66,7 +66,7 @@
 #include "flexmember.h"
 
 #ifndef FALLTHROUGH
-# if __GNUC__ < 7
+# if __GNUC__ < 7 || !__has_attribute(__fallthrough__)
 #  define FALLTHROUGH ((void) 0)
 # else
 #  define FALLTHROUGH __attribute__ ((__fallthrough__))
diff -Naur grep-3.2/lib/fts.c grep-3.2.tpg/lib/fts.c
--- grep-3.2/lib/fts.c	2018-12-15 18:14:26.000000000 +0000
+++ grep-3.2.tpg/lib/fts.c	2018-12-20 21:06:01.835939123 +0000
@@ -200,7 +200,7 @@
 #endif
 
 #ifndef FALLTHROUGH
-# if __GNUC__ < 7
+# if __GNUC__ < 7 || !__has_attribute(__fallthrough__)
 #  define FALLTHROUGH ((void) 0)
 # else
 #  define FALLTHROUGH __attribute__ ((__fallthrough__))
diff -Naur grep-3.2/lib/intprops.h grep-3.2.tpg/lib/intprops.h
--- grep-3.2/lib/intprops.h	2018-09-19 05:05:05.000000000 +0000
+++ grep-3.2.tpg/lib/intprops.h	2018-12-20 21:10:59.278935010 +0000
@@ -227,14 +227,11 @@
 # define _GL_HAS_BUILTIN_OVERFLOW 0
 #endif
 
-/* True if __builtin_add_overflow_p (A, B, C) works.  */
-#define _GL_HAS_BUILTIN_OVERFLOW_P (7 <= __GNUC__)
-
 /* The _GL*_OVERFLOW macros have the same restrictions as the
    *_RANGE_OVERFLOW macros, except that they do not assume that operands
    (e.g., A and B) have the same type as MIN and MAX.  Instead, they assume
    that the result (e.g., A + B) has that type.  */
-#if _GL_HAS_BUILTIN_OVERFLOW_P
+#ifdef HAVE_BUILTIN_MUL_OVERFLOW_P
 # define _GL_ADD_OVERFLOW(a, b, min, max)                               \
    __builtin_add_overflow_p (a, b, (__typeof__ ((a) + (b))) 0)
 # define _GL_SUBTRACT_OVERFLOW(a, b, min, max)                          \
diff -Naur grep-3.2/lib/mbrtowc.c grep-3.2.tpg/lib/mbrtowc.c
--- grep-3.2/lib/mbrtowc.c	2018-09-19 05:04:26.000000000 +0000
+++ grep-3.2.tpg/lib/mbrtowc.c	2018-12-20 21:07:13.002938139 +0000
@@ -36,7 +36,7 @@
 # include "verify.h"
 
 # ifndef FALLTHROUGH
-#  if __GNUC__ < 7
+# if __GNUC__ < 7 || !__has_attribute(__fallthrough__)
 #   define FALLTHROUGH ((void) 0)
 #  else
 #   define FALLTHROUGH __attribute__ ((__fallthrough__))
diff -Naur grep-3.2/lib/quotearg.c grep-3.2.tpg/lib/quotearg.c
--- grep-3.2/lib/quotearg.c	2018-01-07 00:07:08.000000000 +0000
+++ grep-3.2.tpg/lib/quotearg.c	2018-12-20 21:06:43.871938541 +0000
@@ -55,7 +55,7 @@
 #define INT_BITS (sizeof (int) * CHAR_BIT)
 
 #ifndef FALLTHROUGH
-# if __GNUC__ < 7
+# if __GNUC__ < 7 || !__has_attribute(__fallthrough__)
 #  define FALLTHROUGH ((void) 0)
 # else
 #  define FALLTHROUGH __attribute__ ((__fallthrough__))
diff -Naur grep-3.2/lib/unistr/u8-uctomb-aux.c grep-3.2.tpg/lib/unistr/u8-uctomb-aux.c
--- grep-3.2/lib/unistr/u8-uctomb-aux.c	2018-01-07 00:08:03.000000000 +0000
+++ grep-3.2.tpg/lib/unistr/u8-uctomb-aux.c	2018-12-20 21:07:46.453937676 +0000
@@ -21,7 +21,7 @@
 #include "unistr.h"
 
 #ifndef FALLTHROUGH
-# if __GNUC__ < 7
+# if __GNUC__ < 7 || !__has_attribute(__fallthrough__)
 #  define FALLTHROUGH ((void) 0)
 # else
 #  define FALLTHROUGH __attribute__ ((__fallthrough__))
diff -Naur grep-3.2/lib/xalloc-oversized.h grep-3.2.tpg/lib/xalloc-oversized.h
--- grep-3.2/lib/xalloc-oversized.h	2018-01-07 00:07:08.000000000 +0000
+++ grep-3.2.tpg/lib/xalloc-oversized.h	2018-12-20 21:08:33.047937032 +0000
@@ -41,7 +41,7 @@
    positive and N must be nonnegative.  This is a macro, not a
    function, so that it works correctly even when SIZE_MAX < N.  */
 
-#if 7 <= __GNUC__
+#ifdef HAVE_BUILTIN_MUL_OVERFLOW_P
 # define xalloc_oversized(n, s) \
    __builtin_mul_overflow_p (n, s, (__xalloc_count_type) 1)
 #elif 5 <= __GNUC__ && !defined __ICC && !__STRICT_ANSI__
diff -Naur grep-3.2/src/system.h grep-3.2.tpg/src/system.h
--- grep-3.2/src/system.h	2018-01-07 00:09:11.000000000 +0000
+++ grep-3.2.tpg/src/system.h	2018-12-20 21:09:40.504936099 +0000
@@ -108,7 +108,7 @@
 #endif
 
 #ifndef FALLTHROUGH
-# if __GNUC__ < 7
+# if __GNUC__ < 7 || !__has_attribute(__fallthrough__)
 #  define FALLTHROUGH ((void) 0)
 # else
 #  define FALLTHROUGH __attribute__ ((__fallthrough__))
