diff -Naur grep-3.7/config.hin grep-3.7.tpg/config.hin
--- grep-3.7/config.hin	2021-08-14 11:36:28.000000000 +0000
+++ grep-3.7.tpg/config.hin	2021-09-23 17:05:42.844887491 +0000
@@ -2031,7 +2031,7 @@
    return pointers that can be freed via 'free'; it can be used
    only after including stdlib.h.  These macros cannot be used on
    inline functions.  */
-#if _GL_GNUC_PREREQ (11, 0)
+#if _GL_GNUC_PREREQ (11, 0) && !defined(__clang__)
 # define _GL_ATTRIBUTE_DEALLOC(f, i) __attribute__ ((__malloc__ (f, i)))
 #else
 # define _GL_ATTRIBUTE_DEALLOC(f, i)
diff -Naur grep-3.7/m4/gnulib-common.m4 grep-3.7.tpg/m4/gnulib-common.m4
--- grep-3.7/m4/gnulib-common.m4	2021-08-08 15:34:35.000000000 +0000
+++ grep-3.7.tpg/m4/gnulib-common.m4	2021-09-23 16:59:25.526639473 +0000
@@ -154,7 +154,7 @@
    return pointers that can be freed via 'free'; it can be used
    only after including stdlib.h.  These macros cannot be used on
    inline functions.  */
-#if _GL_GNUC_PREREQ (11, 0)
+#if _GL_GNUC_PREREQ (11, 0) && !defined(__clang__)
 # define _GL_ATTRIBUTE_DEALLOC(f, i) __attribute__ ((__malloc__ (f, i)))
 #else
 # define _GL_ATTRIBUTE_DEALLOC(f, i)
