diff -up grep-3.9/lib/stdlib.in.h.1~ grep-3.9/lib/stdlib.in.h
--- grep-3.9/lib/stdlib.in.h.1~	2023-02-26 18:49:48.000000000 +0100
+++ grep-3.9/lib/stdlib.in.h	2023-03-05 22:05:39.820220676 +0100
@@ -103,7 +103,7 @@ struct random_data
    that can be freed by passing them as the Ith argument to the
    function F.  */
 #ifndef _GL_ATTRIBUTE_DEALLOC
-# if __GNUC__ >= 11
+# if __GNUC__ >= 11 && !defined(__clang__)
 #  define _GL_ATTRIBUTE_DEALLOC(f, i) __attribute__ ((__malloc__ (f, i)))
 # else
 #  define _GL_ATTRIBUTE_DEALLOC(f, i)
diff -up grep-3.9/m4/gnulib-common.m4.1~ grep-3.9/m4/gnulib-common.m4
--- grep-3.9/m4/gnulib-common.m4.1~	2023-03-05 22:05:39.820220676 +0100
+++ grep-3.9/m4/gnulib-common.m4	2023-03-05 22:20:56.780700787 +0100
@@ -194,7 +194,7 @@ AC_DEFUN([gl_COMMON_BODY], [
    can be freed via 'free'; it can be used only after declaring 'free'.  */
 /* Applies to: functions.  Cannot be used on inline functions.  */
 #ifndef _GL_ATTRIBUTE_DEALLOC
-# if _GL_GNUC_PREREQ (11, 0)
+# if _GL_GNUC_PREREQ (11, 0) && !defined(__clang__)
 #  define _GL_ATTRIBUTE_DEALLOC(f, i) __attribute__ ((__malloc__ (f, i)))
 # else
 #  define _GL_ATTRIBUTE_DEALLOC(f, i)
